src/cdomains/apron/apronDomain.apron.ml

```diff
@@ -732,23 +453,46 @@ struct
     let texpr1 = Texpr1.of_expr (A.env nd) (Var v') in
     A.substitute_texpr_with Man.mgr nd v texpr1 None
 
-  let meet_tcons d tcons1 =
+  let meet_tcons d tcons1 e =
     let earray = Tcons1.array_make (A.env d) 1 in
     Tcons1.array_set earray 0 tcons1;
-    A.meet_tcons_array Man.mgr d earray
+    let res = A.meet_tcons_array Man.mgr d earray in
+    match Man.name () with
+    | "ApronAffEq" ->
+      let overflow_res res = if IntDomain.should_ignore_overflow (Cilfacade.get_ikind_exp e) then res else d in
```

Comment:
Why is this handled specially?

Reply:
I think is so apron `affeq` and ours behave similarly, such that one can make a fair comparison. The equivalent is done in `meet_tcons` here:

https://github.com/MartinWehking/analyzer/blob/0ab9127c45235a9cf3d32f8af9f8431cc60d9e58/src/cdomains/apron/affineEqualityDomain.apron.ml#L527

Reply:
I guess in the thesis this is described as "positive guard refinement for equality constraints in the context of
overflows", but here is no check for the constraint being equality.

Besides that, I don't really understand what this is achieving even, because 57-affeq/09-bounds_guards_no_ov and 57-affeq/10-bounds_guards.ov pass just as well without it.
Moreover, it's computing bounds of the single variable occurring in the constraint, but in affeq the lower and upper bound must be equal to begin with.
And then refinement is done regardless how the single variable occurs in the constraint. For example `x - 1 == 0` is refined in the same way as `x - 42 == 0`, which seems odd.

The mathematical description

$$
[x]^\sharp D_{refined} \in ikind(x)\text{ with }ikind : Vars \to \mathbb{N} \times \mathbb{N}
$$

in the thesis doesn't make sense to me either. $ikind(x)$ is a pair of naturals (or should it be integers?), but pairs cannot contain ( $\in$ ) some abstract value.

Reply:
> Besides that, I don't really understand what this is achieving even, because 57-affeq/09-bounds_guards_no_ov and 57-affeq/10-bounds_guards.ov pass just as well without it.

Both of those are tests for our `affeq` and not the affine equalities that come with apron though.

Reply:
I mean disabling this special logic in `AffineEqualityDomain` will still pass those.

Reply:
I think the notation is a bit sloppy, it should actually be $\subseteq$ and `ikind` should be from `Vars to V^\sharp` or something, but I think the intent is still clear.

Reply:
> but here is no check for the constraint being equality. 

I think this refers to the constraint inside the domain being equality between  a variable and a constant?

Reply:
> Additionally, we now also refine after positive guards when overflows might occur and there is only one variable inside the expression and the expression is an equality constraint check (==).  We check after the refinement if the new value of the variable is outside its integer bounds and if that is the case, either revert to the old state or set it to bottom.

I think this comment explains the intention here.

Reply:
Also, if I disable these extra refinements, for me test 57/02 crashes.

Reply:
> I think this refers to the constraint inside the domain being equality between a variable and a constant?

It doesn't seem like that though: the affeq version of this explicitly checks for the refining constraint being equality/non-equality.

> > Additionally, we now also refine after positive guards when overflows might occur and there is only one variable inside the expression and the expression is an equality constraint check (==).  We check after the refinement if the new value of the variable is outside its integer bounds and if that is the case, either revert to the old state or set it to bottom.
> 
> I think this comment explains the intention here.

It doesn't make it any clearer to me. That says "after the refinement", but this logic _is_ the refinement.
Also if it's just about variables having out of ikind bounds integer values, what does it have to do with refining _equality_ constraints and why just one variable. That seems like a general reduction step for all variables in the environment.

> Also, if I disable these extra refinements, for me test 57/02 crashes.

That test crashing doesn't explain this refinement though. If this is removed, it should just be less precise and fail some assertion, not crash altogether. If it's just to avoid that crash, this is just a workaround for a crash (due to some particular form of constraint for some reason), not a refinement.

---

Is there an example that directly shows the benefit of this refinement? As in, without this the assertion is unknown, but with this it succeeds.

Reply:
~~~C
int main() {
    int x;
    int itv;
    int top;

    itv = 0;

    if(top) {
      itv = 10;
    }

    if(x == itv) {
      __goblint_check(x == itv);
    }
}
~~~

This is the sort of thing one would hope that this can handle, Unfortunately, it does not seem like it actually helps here.

Reply:
I have identified cases where it matters though. Consider, e.g., the following program


~~~C
//SKIP PARAM: --set ana.activated[+] affeq --set sem.int.signed_overflow assume_top --enable ana.int.interval

int main() {
    int top;

    if(top == 8) {
      __goblint_check(top == 8);
    }
}
~~~

With this refinement, by inspecting the HTMl output, we can see that affeq knows that `top-8=0`, without it, it only has `T`. Though I have not succeeded making an example where this shows in terms of asserts.

Reply:
~~~C
//SKIP PARAM: --set ana.activated[+] affeq --set sem.int.signed_overflow assume_top --enable ana.int.interval

int main() {
    int top;
    int blarg = top;

    if(top == 8) {
      __goblint_check(top == 8);
      __goblint_check(top == blarg);
    }
}
~~~

Took some further constructing, but this example requires the refinement to work.

Reply:
Thanks for constructing the example! I'll add that as a test here.
I originally wanted to extract this common code, but ended up with lesser refactoring since that would've required another functor.

I'm still a bit puzzled by this special logic, but I'll just leave it.
There's no operation that could even overflow here, so refining with `top == 8` should succeed regardless of signedness or overflow settings.

Reply:
The example where things overflow is checks such as `if(x-2 == MAX_INT)` and such, where it would be wrong to add `x = MAX_INT + 2` as s constraint, as that involves a signed wraparound.

Reply:
I'm not sure checking the single variable occurring in the condition is sufficient though.
For example, consider the affeq state `x + y == MAX_INT`  refined by the condition `y == -1`. By Gaussian elimination, this will also give `x == MAX_INT + 1`, but just checking the bounds of the single variable `y` in the condition doesn't catch that.